homeassistant/components/turn_touch.py

```diff
@@ -0,0 +1,178 @@
+"""
+Integrate the Turn Touch smart home remote with Home Assistant.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/turn_touch/
+"""
+
+import logging
+import threading
+import time
+import voluptuous as vol
+
+from homeassistant.const import CONF_DEVICES, CONF_MAC
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.discovery import load_platform
+
+REQUIREMENTS = ['turntouch==0.4.2']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_DEBOUNCE = 'debounce'
+
+DOMAIN = 'turn_touch'
+DATA_KEY = 'DATA_{component}'.format(component=DOMAIN)
+
+DEVICE_SCHEMA = vol.Schema({
+    vol.Required(CONF_MAC): cv.string,
+    vol.Optional(CONF_DEBOUNCE, default=True): cv.boolean,
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_DEVICES, default=[]): [DEVICE_SCHEMA],
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+EVENT_NAME = 'turntouch_press'
+EVENT_DATA_ADDRESS = 'remote_address'
+EVENT_DATA_ENTITY_ID = 'entity_id'
+EVENT_DATA_NAME = 'remote_name'
+EVENT_DATA_TYPE = 'press_type'
+
+RETRY_DELAY = .25
+
+TURN_TOUCH_PLATFORMS = ['sensor']
+
+
+def setup(hass, config):
+    """
+    Set up the Turn Touch platform.
+
+    Also, set up other components (eg. sensor) provided by this platform.
+    """
+    hass.data[DATA_KEY] = {'devices': {}}
+
+    for device_config in config[DOMAIN][CONF_DEVICES]:
+        turn_touch = TurnTouchRemote(hass, device_config)
```

Comment:
Entity instances should not be created by the component but by the platform. Move this to the platform.

Reply:
Thanks for the feedback @MartinHjelmare. I'm a bit confused about how this would work, and I'd appreciate some more guidance.

I initialize the device connection from the component because this is a bluetooth device and we can only have a single connection to it. If the device were (hypothetically) extended to support more than one type of sensor (lets say battery sensor + humidity sensor), how would that work if the entity is defined/initialized in each of the platforms?

Right now there is only one platform (the battery sensor), but even so, the real/primary device is the remote, not the battery sensor. So it seems logical to represent (& connect to) the remote in the component, and use the platform only for the battery level part.

Reply:
Separate the entity from the connection and keep the connection in the component and the entity creation in each platform.